tools/codegen/api/cpp.py

```diff
@@ -191,9 +195,89 @@ def argument(a: Union[Argument, TensorOptionsArguments, ThisArgument]) -> CppArg
     else:
         assert_never(a)
 
-def group_arguments(
-    func: FunctionSchema, *, method: bool = False
-) -> Sequence[Union[Argument, TensorOptionsArguments, ThisArgument]]:
+@dataclass(frozen=True)
+class CppSignature:
+    arguments: Sequence[Union[Argument, TensorOptionsArguments, ThisArgument]]
+
+    def cpp_arguments(self) -> Sequence[CppArgument]:
+        return list(map(argument, self.arguments))
+
+    # Return arguments as a comma separated list, i.e. like they would be in a C++
+    # function signature. Include default values for arguments.
+    def cpp_arguments_str(self) -> str:
+        return ', '.join(map(str, self.cpp_arguments()))
+
+    # Return a copy of the CppSignature with the `this` argument removed
+    def exclude_this(self) -> 'CppSignature':
```

Comment:
nit: I probably wouldn't enshrine this into the data model (I understand this is refactored from some code that used to live in gen.) The reasoning is that with `exclude_this`, I now have to be aware if a `CppSignature` that I have in hand is either a full CppSignature (with everything) or a "stripped" version (post calling `exclude_this`). But you always *immediately* call some sort of rendering function (e.g., `cpp_arguments_str`) afterwards. We should make the illegal state (a free floating stripped `CppSignature` object) unrepresentable, and we can do this by pushing the defaulting and exclude this behavior directly into kwargs for the relevant rendering functions. WDYT?

Reply:
Though, it turns out I may have more fundamental problems with this class. Check https://github.com/pytorch/pytorch/pull/44087/files#r496137831

Reply:
I made it into a parameter